src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -341,6 +381,706 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
 
 
 /*
+ * AllRelationsJoinedOnPartitionKey aims to deduce whether each of the RTE_RELATION
+ * is joined with at least on another RTE_RELATION on their partition keys. If each
+ * RTE_RELATION follows the above rule, we can conclude that all RTE_RELATIONs are
+ * joined on their partition keys.
+ *
+ * In order to do that, we invented a new equivalence class namely:
+ * AttributeEquivalenceClass. In very simple words, a AttributeEquivalenceClass is
+ * identified by an unique id and consists of a list of AttributeEquivalenceMembers.
+ *
+ * Each AttributeEquivalenceMember is designed to identify attributes uniquely within the
+ * whole query. The necessity of this arise since varno attributes are defined within
+ * a single level of a query. Instead, here we want to identify each RTE_RELATION uniquely
+ * and try to find equality among each RTE_RELATION's partition key.
+ *
+ * Each equality among RTE_RELATION is saved using an AttributeEquivalenceClass where
+ * each member attribute is identified by a AttributeEquivalenceMember. In the final
+ * step, we try generate a common attribute equivalence class that holds as much as
+ * AttributeEquivalenceMembers whose attributes are a partition keys.
+ *
+ * AllRelationsJoinedOnPartitionKey uses both relation restrictions and join restrictions
+ * to find as much as information that Postgres planner provides to extensions. For the
+ * details of the usage, please see GenerateAttributeEquivalencesForRelationRestrictions()
+ * and GenerateAttributeEquivalencesForJoinRestrictions()
+ *
+ * Finally, as the name of the function reveals, the function returns true if all relations
+ * are joined on their partition keys. Otherwise, the function returns false.
+ */
```

Comment:
> This doesn't actually tests node restrictions, so that'd not be any better?

Yes, this function is not doing any shard pruning or adding restrictions. We use this function as a helper function to decide whether to add or not to add the restrictions.

Some other suggestions:

* `SafeToPushDownSubquery()` (actually, in the `Union`s PR, I've switched to this name)
* `AddShardIntervalRestrictionsToRelations()`
* `AddShardIntervalRestrictions()`

Any more suggestions?